#Fluxos Monetários

Já sabemos que o dinheiro apresenta variação em seu valor ao longo do tempo. Essa premissa é bastante utilizada para cálculo de pagamentos futuros e contínuos. Pagamentos contínuos são
formalmente conhecidos como **Séries Uniformes**. 

Séries uniformes são uma sequência de pagamentos (ou recebimentos) de valor constante ao longo do tempo em fluxo de caixa.


Em Python, existe uma biblioteca *Numpy-Financial* que apresenta várias [funções financeiras](https://github.com/levisouuza/numpy-financial), inclusive, a de séries de pagamentos, chamada **pmt**. Para utilizá-la usaremos:

In [1]:
!pip install numpy-financial

In [3]:
#numpy.pmt(rate, nper, pv, fv=0, when='end')
#rate: Taxa de juros proporcionalmente ajustado ao período de análise.
#nper: Número de períodos
#pv: Present value ou Valor presente líquido. Quantidade inicial.
#fv: Future Value ou valor futuro/valor final.

#DISCLAIMER: A biblioteca foi descontinuada, mas neste momento, 03/06/2020, ainda é possível utilizá-la.  Além disso, também será criada uma função para calcular
#a quantidade de pagamentos.

import numpy_financial as np
from math import pow


**Exemplo 1**: Uma loja virtual uma TEV LED 40", por R$ 1.699,00 à vista ou em 10x, cobrando uma taxa de juros de 1,79% a.m. Qual o valor da parcela ? 

In [4]:
#utilizando a biblioteca numpy
pgto = np.pmt(0.0179, 10, 1699)

In [5]:
round(pgto,2)

-187.07

In [ ]:
#utilizando função própria.
def pagamentos(rate, nper, pv, fv=0):
  pgto = pv*((rate*pow((1+rate),nper))/(pow((1+rate),nper)-1))
  return round(pgto,2)*-1

In [ ]:
pagamentos(0.0179, 10, 1699)

-187.07

Nesse caso, o valor do pagamento deverá ser negativo, visto que é uma saída do fluxo de caixa. 

Também existem as séries uniformes postecipadas que consideram o valor de pagamento àquele no instante n = 1.

**Exemplo 2**: Foi realizada a compra de um eletrodoméstico a prazo, 4 parcelas de R$150,00. Sabendo que a loja cobra taxa de juros de 7.714% a.m, qual seria o valor do pagamento à vista ?

É importante destacar o conceito de valor presente ou valor presente líquido (VPL). Valor presente é o valor pago por determinado investimento no instante t0 de um fluxo de caixa. Os valores futuros são corrigidos pela taxa de juros em questão.

Nos casos de valores presentes de uma série de pagamentos postecipadas, a expressão que implementaremos é: **COLOCAR EQUACAO**.

In [6]:
#Implementando resultado

def pv_postecipada(rate, nper, pmt):
  pv = pmt*((pow((1+rate),nper) - 1) / (pow((1+rate),nper)*rate))
  return round(pv,2)

In [7]:
pv_postecipada(0.07714,4,150)

500.0

**Exemplo 3**: Um jovem de 20 anos quer investir R$200,00/mês até os seus 60 anos. A partir dessa idade até seus 85 anos, qual o valor do crédito mensal em sua conta corrente, supondo-se que a remuneração dos recursos acumulados será constante e igual a inflação +0.30% a.m.

In [ ]:
#Inicialmente, calcularemos o futuro da série de pagamentos.

def fv_aposentadoria(rate, nper, pmt):
  fv = pmt*((pow((1+rate),nper)-1)/rate)
  return round(fv,2)

In [ ]:
#tempo de contribuição será igual a 60-20 = 40. Como a taxa está descrita de forma mensal, iremos transformar 40 anos em meses, onde teremos 480 meses.
valor_futuro = fv_aposentadoria(0.0030, 480, 200)
valor_futuro

214107.14

In [ ]:
#tempo de benefício: 85-60 = 25 = 300 meses. 
pagamentos(0.0030, 300, -valor_futuro)

1083.39

###Perpetuidade



Muitas aplicações financeiras apresentam duração e rentabilidade bem definidas no momento de sua aquisição, são os casos de investimentos em renda fixa. 

Mas, e para aqueles investimentos que não possuem "data de validade" ? Como descobrir o valor presente de um investimento "eterno" ?

Para esse tipo de problemas, é aplicado o conceito de *Perpetuidade* ou *Valor presente de Perpetuidade*. A perpetuidade consiste em informar o valor presente de aplicações financeiras que apresentam fluxos de caixas contínuos em um período indeterminado.

Esse método é amplamente utilizado no modelo de precificação de empresas, *valuation*, que tem por objetivo apresentar o valor de mercado das instituições baseados nos seus fluxos de caixa futuros a valor presente.

**Exemplo 4**: Uma empresa de saneamento pagou dividendos por ações no valor de R$ 2,73 em 2012. Seus lucros e dividendos cresceram a uma taxa de 6% a.a entre 2000 e 2012, e pressupõe-se que cresçam a essa mesma taxa em anos futuros. Sabe-se que a taxa de aplicação(tesouro) requerida pelos investidores está em 12,23%a.a.Qual deveria ser valor de uma ação dessa empresa ?

In [ ]:
#Utilizaremos a função de   perpetuidade

def perpetuidade(rate, pmt, growth):
  pv = pmt*((1+growth)/(rate-growth))
  return round(pv,2)

In [ ]:
perpetuidade(0.1223, 2.73, 0.06)

46.45

Logo, o valor justo da ação dessa empresa seria de R$46.45.